doc/lsst.pipe.base/creating-a-command-line-task.rst

```diff
@@ -0,0 +1,253 @@
+If this is not the case you can easily provide a modified argument parser.
+
+Typically this consists of constructing an instance of `lsst.pipe.base.ArgumentParser` and then adding some ID arguments to it using `~lsst.pipe.base.argumentParser.ArgumentParser.add_id_argument`.
+This is shown in several examples below.
+Please resist the urge to add other kinds of arguments to the argument parser unless truly needed.
+One strength of our tasks is how similar they are to each other.
+Learning one set of arguments suffices to use many tasks.
+
+.. warning::
+
+   If your task requires a custom argument parser to do more than just change the type of the single data reference, then it also require a :ref:`custom task runner <creating-a-command-line-task-custom-task-runner>` as well.
+
+Here are some examples:
+
+- A task's ``run`` method requires a data reference of some kind other than a raw or calibrated image.
+  This is a common case, and easily solved.
+  For example the ``processCoadd.ProcessCoaddTask`` processes co-adds, which are specified by sky map patch.
+  Here is ``ProcessCoaddTask._makeArgumentParser``:
+
+  .. code-block:: text
+
+     @classmethod
+     def _makeArgumentParser(cls):
+         """Create an argument parser
+         """
+         parser = pipeBase.ArgumentParser(name=cls._DefaultName)
+         parser.add_id_argument("--id", "deepCoadd", help="data ID, e.g. --id tract=12345 patch=1,2",
+                                ContainerClass=CoaddDataIdContainer)
+         parser.add_id_argument("--selectId", "calexp", help="data ID, e.g. --selectId visit=6789 ccd=0..9",
+                                ContainerClass=SelectDataIdContainer)
+         return parser
+
+  - The first argument to `~lsst.pipe.base.argumentParser.ArgumentParser` is the name of the ID argument.
+  - The second argument is a dataset type, which specifies the keys that are used with this ID argument.
+    The keys associated with a particular dataset type are specified in the mapper configuration file for the observatory package and camera in question, and thus may vary from camera to camera.
+    In practice, the keys for ``raw`` and ``calexp`` dataset types usually do vary from camera to camera, but the keys for coadds do not.
+    However, this is not a fixed rule.
+    For most observatory packages ``deepCoadd`` is one of two coadd dataset types, and the other, ``goodSeeingCoadd``, would work just as well for this argument.
+  - A custom ``ContainerClass`` (for example, `lsst.coadd.utils.coaddDataIdContainer.CoaddDataIdContainer`) is provided to support iterating over missing keys (e.g. if you provide a tract but not a patch then the task will iterate over all available patches for that tract).
+    This happens automatically for ``raw`` and ``calexp`` dataset types, but not most other dataset types.
+    Examine the code in `~lsst.coadd.utils.coaddDataIdContainer.CoaddDataIdContainer` to see how it works.
```

Comment:
Should they be reading this code or should we fix/generalise it?  I think the latter.